rfcs/taskforce/0002-network-utility.md

```diff
@@ -0,0 +1,108 @@
+5. We have a variety of charts and metrics for monitoring macro network health but debugging specific issues is often manual
+
+### Improvements
+
+While this system is effective, there are a variety of potential improvements to this process.
+
+- Keypair generation depends on the daemon
+- Genesis ledger logic could be cleaned up and made more flexible to support a wider variety of configuraitons
+- These tools could be packaged together with a JavaScript API for extending functionality
+- The internal data structures for storing these keypairs, genesis configuration and distribution are not well defined and could be more robust
+- Tools for deployment and monitoring are seperate and these could be packaged in one simple workflow that uses the same underlying functionality
+
+## Proposal
+
+All of these features should be included in a simple command line tool that also exposes a TypeScript API. This service will be written in Reason with TypeScript bindings.
+
+### Keypair Management
+
+For new keypair generation, we can use the new client-sdk which is a much lighter dependency and has a clean JavaScript interface.
+
+```
+type keypair = {
+  publicKey: string,
+  privateKey: option(string)
+}
+```
+
+However there are also existing keys which we'll need to be able to reference by public key alone since we don't control the private keys. These include community owned wallets for inclusion in the genesis ledger.The basic type of a keypair consists of the following:
+
+### Keyset Management
+
+Keysets are simply a collection of keypairs. We use these for categorizing different groups of keys that will be used in the lifecycle of a network. For each keypair in a keyset, we MAY or MAY NOT have the corresponding private keys. For each keyset, we may want to allocate a different amount of funds to their corresponding accounts in a genesis ledger.
+
+```
+type keysetMember = | Owned(keypair) | Unowned(publicKey)
+```
+
+Where it's implied that `type publicKey = string`.
+
+### Runtime Ledger Generation
+
+The code for generating the Runtime Genesis Ledger should be ported over from the [existing code](https://github.com/CodaProtocol/coda-automation/blob/master/scripts/testnet-keys.py#L203) but with the added flexibility of being able to inject any variety of keysets into the ledger with corresponding initial amounts.
+
+### Configuration Deployment
+
+The configuration deployment should be updated to take advantage of the kubernetes JavaScript client API. This tool should expose similar functionality to the existing tool, allowing users to upload keypairs to the kubernetes ConfigMap however not splitting out the task into predefined groups but rather allowing arbitrary keysets to be deployed.
+
+### Network Deployment and Monitoring
+
+This tool could optionally wrap the coda-automation tools and provide a simple interface for deploying a full testnet from a single command.
```

Comment:
Maybe keep this out-of-scope for the initial implementation.

Though, this may be the right place to include the "start/stop node" functionality we'll need for our integration tests. We'll keep in contact as we get closer to working on that part of it.

Reply:
I have a hacked together shell script which sort of provides a workflow you can lean off of for doing this. Probably should be lifted to something that is not a shell script if it needs to be changed significantly, but it's basically just a pipeline of commands to run. Checkout `scripts/auto-deploy.sh` in `coda-automation`.

Reply:
Yeah that looks good, not sure if we even need to deprecate that logic for now, can just call into it using this tool.